File: qt/gen_qtipccommon.cpp
// +build ignore

#include <QNativeIpcKey>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qtipccommon.h>
#include "gen_qtipccommon.h"

#ifndef _Bool
#define _Bool bool
#endif

void _GUID_Delete(_GUID* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<_GUID*>( self );
	} else {
		delete self;
	}
}

void type_info_Delete(type_info* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<type_info*>( self );
	} else {
		delete self;
	}
}

QNativeIpcKey* QNativeIpcKey_new() {
	return new QNativeIpcKey();
}

QNativeIpcKey* QNativeIpcKey_new2(Type typeVal) {
	return new QNativeIpcKey(typeVal);
}

QNativeIpcKey* QNativeIpcKey_new3(struct miqt_string k) {
	QString k_QString = QString::fromUtf8(k.data, k.len);
	return new QNativeIpcKey(k_QString);
}

QNativeIpcKey* QNativeIpcKey_new4(QNativeIpcKey* other) {
	return new QNativeIpcKey(*other);
}

QNativeIpcKey* QNativeIpcKey_new5(struct miqt_string k, Type typeVal) {
	QString k_QString = QString::fromUtf8(k.data, k.len);
	return new QNativeIpcKey(k_QString, typeVal);
}

Type QNativeIpcKey_LegacyDefaultTypeForOs() {
	return QNativeIpcKey::legacyDefaultTypeForOs();
}

void QNativeIpcKey_OperatorAssign(QNativeIpcKey* self, QNativeIpcKey* other) {
	self->operator=(*other);
}

void QNativeIpcKey_Swap(QNativeIpcKey* self, QNativeIpcKey* other) {
	self->swap(*other);
}

bool QNativeIpcKey_IsEmpty(const QNativeIpcKey* self) {
	return self->isEmpty();
}

bool QNativeIpcKey_IsValid(const QNativeIpcKey* self) {
	return self->isValid();
}

Type QNativeIpcKey_Type(const QNativeIpcKey* self) {
	return self->type();
}

void QNativeIpcKey_SetType(QNativeIpcKey* self, Type typeVal) {
	self->setType(typeVal);
}

struct miqt_string QNativeIpcKey_NativeKey(const QNativeIpcKey* self) {
	QString _ret = self->nativeKey();
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

void QNativeIpcKey_SetNativeKey(QNativeIpcKey* self, struct miqt_string newKey) {
	QString newKey_QString = QString::fromUtf8(newKey.data, newKey.len);
	self->setNativeKey(newKey_QString);
}

struct miqt_string QNativeIpcKey_ToString(const QNativeIpcKey* self) {
	QString _ret = self->toString();
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}

QNativeIpcKey* QNativeIpcKey_FromString(struct miqt_string stringVal) {
	QString stringVal_QString = QString::fromUtf8(stringVal.data, stringVal.len);
	return new QNativeIpcKey(QNativeIpcKey::fromString(stringVal_QString));
}

void QNativeIpcKey_Delete(QNativeIpcKey* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<QNativeIpcKey*>( self );
	} else {
		delete self;
	}
}

File: qt/gen_qtipccommon.go
package qt

import (
	"unsafe"
)

type QNativeIpcKey__Type uint16

const (
	QNativeIpcKey__SystemV       QNativeIpcKey__Type = 81
	QNativeIpcKey__PosixRealtime QNativeIpcKey__Type = 256
	QNativeIpcKey__Windows       QNativeIpcKey__Type = 257
)

type QNativeIpcKey struct {
	h          uintptr
	isSubclass bool
}

// NewQNativeIpcKey constructs a new QNativeIpcKey object.
func NewQNativeIpcKey() *QNativeIpcKey {
	ret := newQNativeIpcKey(QNativeIpcKey_new())
	ret.isSubclass = true
	return ret
}

// NewQNativeIpcKey2 constructs a new QNativeIpcKey object.
func NewQNativeIpcKey2(typeVal Type) *QNativeIpcKey {
	ret := newQNativeIpcKey(QNativeIpcKey_new2(typeVal))
	ret.isSubclass = true
	return ret
}

// NewQNativeIpcKey3 constructs a new QNativeIpcKey object.
func NewQNativeIpcKey3(k string) *QNativeIpcKey {
	k_ms := struct_miqt_string{}
	k_ms.data = CString(k)
	k_ms.len = size_t(len(k))
	defer free(unsafe.Pointer(k_ms.data))

	ret := newQNativeIpcKey(QNativeIpcKey_new3(k_ms))
	ret.isSubclass = true
	return ret
}

// NewQNativeIpcKey4 constructs a new QNativeIpcKey object.
func NewQNativeIpcKey4(other *QNativeIpcKey) *QNativeIpcKey {
	ret := newQNativeIpcKey(QNativeIpcKey_new4(other.cPointer()))
	ret.isSubclass = true
	return ret
}

// NewQNativeIpcKey5 constructs a new QNativeIpcKey object.
func NewQNativeIpcKey5(k string, typeVal Type) *QNativeIpcKey {
	k_ms := struct_miqt_string{}
	k_ms.data = CString(k)
	k_ms.len = size_t(len(k))
	defer free(unsafe.Pointer(k_ms.data))

	ret := newQNativeIpcKey(QNativeIpcKey_new5(k_ms, typeVal))
	ret.isSubclass = true
	return ret
}

func QNativeIpcKey_LegacyDefaultTypeForOs() Type {
	xxxxxxxxx
}

func (this *QNativeIpcKey) OperatorAssign(other *QNativeIpcKey) {
	QNativeIpcKey_OperatorAssign(this.h, other.cPointer())
}

func (this *QNativeIpcKey) Swap(other *QNativeIpcKey) {
	QNativeIpcKey_Swap(this.h, other.cPointer())
}

func (this *QNativeIpcKey) IsEmpty() bool {
	return (bool)(QNativeIpcKey_IsEmpty(this.h))
}

func (this *QNativeIpcKey) IsValid() bool {
	return (bool)(QNativeIpcKey_IsValid(this.h))
}

func (this *QNativeIpcKey) Type() Type {
	xxxxxxxxx
}

func (this *QNativeIpcKey) SetType(typeVal Type) {
	QNativeIpcKey_SetType(this.h, typeVal)
}

func (this *QNativeIpcKey) NativeKey() string {
	var _ms struct_miqt_string = QNativeIpcKey_NativeKey(this.h)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QNativeIpcKey) SetNativeKey(newKey string) {
	newKey_ms := struct_miqt_string{}
	newKey_ms.data = CString(newKey)
	newKey_ms.len = size_t(len(newKey))
	defer free(unsafe.Pointer(newKey_ms.data))
	QNativeIpcKey_SetNativeKey(this.h, newKey_ms)
}

func (this *QNativeIpcKey) ToString() string {
	var _ms struct_miqt_string = QNativeIpcKey_ToString(this.h)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QNativeIpcKey_FromString(stringVal string) *QNativeIpcKey {
	stringVal_ms := struct_miqt_string{}
	stringVal_ms.data = CString(stringVal)
	stringVal_ms.len = size_t(len(stringVal))
	defer free(unsafe.Pointer(stringVal_ms.data))
	_goptr := newQNativeIpcKey(QNativeIpcKey_FromString(stringVal_ms))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}
